### ISSUE TITLE
CA1062 reporting for protected and internal methods

### ISSUE THREAD

**Issue by lonix1** [dotnet/roslyn-analyzers#5726]
### Analyzer

**Diagnostic ID**: [CA1062](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1062)

CA1062
"Validate method parameter is non-null before using it"

### Describe the improvement

CA1062 is reported here:
````cs
protected virtual int Foo(string s) => s.Length;
````

It is not reported here:
````cs
internal virtual int Bar(string s) => s.Length;
````

This doesn't seem right. The internal method can be accessed by any caller in the assembly, and the analyzer would allow nulls. But it ignores the protected method, which has less callers, which presumably are more careful.

The behavior in these cases should be reconsidered.

I'm not sure what would be best, but I think 1) the behavior should be swapped, or 2) they should both trigger the analyzer.

### Describe suggestions on how to achieve the rule
no

### Additional context
no

**Comment by svick** [comment]
I think the distinction here is that for an `internal` method, you can see all callers and they are also under your control. But a `protected` method can be called from another assembly, and there is no guarantee you can see or control all such assemblies, which means you should check for `null`.

**Comment by lonix1** [comment]
I think you are right. But in that case why not trigger the analyser for both cases?

Within an assembly, an internal method is as good as public. So surely is should also trigger the analyser?

**Comment by svick** [comment]
@lonix1 Basically, the expectation is that code within an assembly works closely together. You're not going to call the method with `null`, because you can see that the implementation does not allow it. And even if you do call it with `null`, it's not going to be hard for you to diagnose the error.

But code outside the assembly may not see the implementation. And if you pass in `null`, you want to have a clear exception that speaks in the language of the API ("parameter `s` is null"), not in the language of the implementation ("something on line 42 is `null`").

That's why boundary that's important here is the assembly, not a single class.

**Comment by lonix1** [comment]
I suppose you're right.

But, I tend to test args for internal as well. I think the boundary is not the assembly, but rather the developer who's using it. In a team when multiple people work on the same assembly, one guy blows up the other guy's code. So I always test internal args as well.

I guess we can close this as "by design"?

**Comment by Evangelink** [comment]
@mavasani It's pretty easy to add support for [api_ surface](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#analyzed-api-surface) configuration for this rule instead of the `IsExternallyVisible`. I don't really see any downside here, WDYT?

**Comment by mavasani** [comment]
@Evangelink Either way is fine with me. If you add the support for the option, please open a doc issue to update the documentation.
